compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -326,6 +324,16 @@ BLangStatementExpression desugar(BLangQueryExpr queryExpr, SymbolEnv env,
         return streamStmtExpr;
     }
 
+    private BType getMapType(BType type) {
+        BType refType = Types.getReferredType(type);
+        if (refType.tag == TypeTags.UNION) {
+            return getMapType(types.getSafeType(refType, false, true));
```

Comment:
getSafeType() handles unions, hence we don't need the union check

Reply:
Addressed in https://github.com/ballerina-platform/ballerina-lang/pull/36743/commits/fb95e5587428225ee1a8284b66e3a2efd22b28c2.